File: C++/Easy/73.cpp
/*
Daily Coding Problem: 73
Difficulty: Easy
Asked by: Google
Authored by: TenthEdict
*/

#include <iostream>

struct Node
{
   int val;
   Node* next;

   Node(int v, Node* n = nullptr) : val(v), next(n) {}
};

Node* reverse(Node* curr) 
{
    Node* prev = nullptr;
    
    while (curr) 
    {
        Node* next = curr->next;
        curr->next = prev;
        prev = curr;
        curr = next;
    }
    return prev;
};

void print_list(Node* head)
{
    while (head)
    {
        std::cout << head->val << std::endl;

        head = head->next;
    }
};

int main()
{
    Node* ptr3 = new Node(3);
    Node* ptr2 = new Node(2, ptr3);
    Node* ptr1 = new Node(1, ptr2);

    print_list(reverse(ptr1));

    return 0;
}

File: TypeScript/Easy/73.ts
/*
 * Daily Coding Problem: 73
 * Difficulty: Easy
 * Asked By: Google
 * Authored By: TenthEdict
 */

interface ListNode {
    val: number;
    next: ListNode | null;
}

function reverse(head: ListNode | null): ListNode | null {
    let prev = null; 
    let curr = head; 

    while (curr) {
        let next = curr.next;
        curr.next = prev;
        prev = curr;
        curr = next
    }
    return prev;
}

// Helper to visualize the list
function printList(head: ListNode | null): void {
    while (head) {
        console.log(head.val);
        head = head.next;
    }
}

// Build: 1 -> 2 -> 3 -> null
const node3: ListNode = {
    val: 3,
    next: null,
}
const node2: ListNode = {
    val: 2,
    next: node3,
}

const node1: ListNode = {
    val: 1,
    next: node2,
}

// Then reverse it and print both

console.log(printList(reverse(node1)));